### DISCUSSION TITLE
Production build: Text `[Object object]` on black background visible before actual page loads

### DISCUSSION THREAD

**Discussion by andrekovac** [gatsbyjs/gatsby#37688] [Help] [answered]
When I initially go to the **production build** (or reload the page in the browser), I see `[Object object]` on a black background for a short time before the actual page becomes visible. I want to get rid of this and directly show the page (as it happens during development)

This is what is visible at first: 

<img width="361" alt="Screenshot 2023-01-19 at 23 45 39" src="https://user-images.githubusercontent.com/1945462/220605284-1a75ba76-75b9-4487-b65e-14de61505152.png">

You can visit [https://www.andrekovac.com/](https://www.andrekovac.com/) to observe this behavior.

What could Gatsby want to display here? `[Object object]` hints to some JS object not being displayed properly. But I don't even know where to start to investigate.. My first guesses would be `gatsby-browser.js`, `gatsby-ssr.js` and `gatsby-node.js`, but they all seem right.

**Notes:** 
- This **only happens in the production build** (created with `gatsby build`) of my gatsby site, not in the development build.
- The more I throttle the internet connection via the browser dev tools, the longer the `[Object object]` text remains visible before the entire page becomes visible.
You also get it when building and running the gatsby site locally via `gatsby build` and then `gatsby serve`.

Gatsby Version: `5.4.2`

Happy to share more details if helpful!

**Comment by gh640** [comment, marked as answer]
The generated HTML file https://www.andrekovac.com/index.html seems to have `[object Object]` in the body.

```html
  <div id="___gatsby">
   [object Object]
  </div>
```

Normally this should be a rendered HTML as you know.

Can you share the following?

- the stdout / stderr when you run the production build
- Which Gatsby API functions you use (like `onRenderBody`, `onPostBuild`)
- which packages you use (the output of `gatsby info`)

**Reply by andrekovac** [reply]
Thanks @gh640 ! Your hints helped me to find the issue. In `gatsby-ssr.js` I had this `replaceRenderer` function:

```js
import i18n from "./src/lib/i18n";

export const replaceRenderer = ({ bodyComponent, replaceBodyHTMLString }) => {
  i18n.loadNamespaces(["common"], () => {
    replaceBodyHTMLString(bodyComponent);
  });
};
```

where `./src/lib/i18n` is a wrapper around `i18n` from "i18next".
I copied the code from https://www.gatsbyjs.com/blog/2017-10-17-building-i18n-with-gatsby/#without-redux . There it even says **"Not yet tested"**. Well, it leads to the given error. 

Currently, I still disable multi-lingual texts, so I will test in the future whether I need this code at all.

---

P.S. Because the first page load time is a bit high, I added a **_Loading..._** `div` to `onRenderBody` (in `gatsby-ssr.js`) to show a loading text until the initial page is shown. It probably takes a bit longer for me because of many images on the page.

**Reply by gh640** [reply]
Thank you for the detailed explanation! Glad to hear you solved the issue :)
